dll/win32/shell32/dialogs/item_prop.cpp

```diff
@@ -0,0 +1,179 @@
+            if (pData->pObjStream)
+                pData->pObjStream->Release();
+            hr = E_FAIL;
+        }
+        if (FAILED(hr))
+            FreeData(pData);
+        return hr;
+    }
+
+    static DWORD CALLBACK ShowPropertiesThread(LPVOID Param)
+    {
+        DATA *pData = (DATA*)Param;
+        CComPtr<IDataObject> pDO;
+        if (pData->pObjStream)
+            CoGetInterfaceAndReleaseStream(pData->pObjStream, IID_PPV_ARG(IDataObject, &pDO));
+        Show(pData->pClsidDefault, pDO, pData->InitFunc, pData->InitString);
+        FreeData(pData);
+        return 0;
+    }
+};
+
+static void CALLBACK
+FSFolderItemPropDialogInitCallback(LPCWSTR InitString, IDataObject *pDO, HKEY *hKeys, UINT *cKeys)
+{
+    UNREFERENCED_PARAMETER(InitString);
+    CDataObjectHIDA cida(pDO);
+    if (SUCCEEDED(cida.hr()) && cida->cidl)
+    {
+        PCUITEMID_CHILD pidl = HIDA_GetPIDLItem(cida, 0);
+        AddFSClassKeysToArray(1, &pidl, hKeys, cKeys); // Add file-type specific pages
+    }
+}
+
+static void CALLBACK
+ClassPropDialogInitCallback(LPCWSTR InitString, IDataObject *pDO, HKEY *hKeys, UINT *cKeys)
+{
+    UNREFERENCED_PARAMETER(pDO);
+    AddClassKeyToArray(InitString, hKeys, cKeys); // Add pages from HKCR\%ProgId% (with shellex\PropertySheetHandlers appended later)
+}
+
+HRESULT
+SHELL32_ShowFilesystemItemPropertiesDialogAsync(IDataObject *pDO)
+{
+    CDataObjectHIDA cida(pDO);
+    HRESULT hr = cida.hr();
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    const CLSID *pClsid = NULL;
+    ShellPropSheetDialog::PFNINITIALIZE InitFunc = NULL;
+    LPCWSTR InitString = NULL;
+
+    if (_ILIsDrive(HIDA_GetPIDLItem(cida, 0)))
+    {
+        pClsid = &CLSID_ShellDrvDefExt;
+        InitFunc = ClassPropDialogInitCallback;
+        InitString = L"Drive";
+    }
+    else
+    {
+        pClsid = &CLSID_ShellFileDefExt;
+        InitFunc = FSFolderItemPropDialogInitCallback;
+    }
+    ShellPropSheetDialog Dialog;
+    return Dialog.ShowAsync(pClsid, pDO, InitFunc, InitString);
+}
+
+HRESULT
+SHELL32_ShowShellExtensionProperties(const CLSID *pClsid, IDataObject *pDO)
+{
+    WCHAR ClassBuf[6 + 38 + 1] = L"CLSID\\";
+    StringFromGUID2(*pClsid, ClassBuf + 6, 38 + 1);
+    ShellPropSheetDialog Dialog;
+    return Dialog.ShowAsync(NULL, pDO, ClassPropDialogInitCallback, ClassBuf);
+}
+
+HRESULT
+SHELL_ShowItemIDListProperties(LPCITEMIDLIST pidl)
+{
+    assert(pidl);
+
+    CComHeapPtr<ITEMIDLIST> alloc;
+    if (IS_INTRESOURCE(pidl))
+    {
+        HRESULT hr = SHGetSpecialFolderLocation(NULL, LOWORD(pidl), const_cast<LPITEMIDLIST*>(&pidl));
+        if (FAILED(hr))
+            return hr;
+        alloc.Attach(const_cast<LPITEMIDLIST>(pidl));
+    }
+    SHELLEXECUTEINFOA sei = {
+        sizeof(sei), SEE_MASK_INVOKEIDLIST | SEE_MASK_ASYNCOK, NULL, "properties",
+        NULL, NULL, NULL, SW_SHOWNORMAL, NULL, const_cast<LPITEMIDLIST>(pidl)
+    };
+    return ShellExecuteExA(&sei) ? S_OK : HResultFromWin32(GetLastError());
```

Comment:
Why not using `ShellExecuteExW` (with `SHELLEXECUTEINFOW`) directly?
Is this because the pidl can be an ansi string?

Reply:
PIDLs can be ANSI or Unicode internally, we don't control that. I used the A function for the verb string. (CDefaultContextMenu.Cpp already has all the known verb strings as PCSTR)